### PR TITLE
Adds Animation to shader vertex graph

### PR DESCRIPTION
**Pull request by Kuranes**
- Optimizes Heavily shader code, reusable BoneMatrix result transform
- Adds Size uniform parameter for Array of uniform from Compiler/Node/shaderGraph
- Sort Shader Attributes inside the DeclareGlobal Compiler code (and beautifies)

### CODE REVIEW

File: sources/osg/Uniform.js [hunk 1]
```diff
@@ -43,14 +43,16 @@ define( [
         },
         apply: function ( gl, location ) {
             if ( this._dirty ) {
-                this.update.call( this.glData, this.data );
+                if ( this.update )
```
Comment (cedricpinson) [inline sources/osg/Uniform.js]: why do you need to test update ?

Comment (Kuranes) [inline sources/osg/Uniform.js]: because when it's a type array there is no update

File: sources/osg/Uniform.js [hunk 2]
```diff
@@ -406,8 +430,10 @@ define( [
         };
         uniform.apply = uniform.applyMatrix;
         uniform.transpose = false;
-        uniform.glData = new MACROUTILS.Float32Array( uniform.data );
-        uniform.update = Uniform.prototype._updateFloat16;
+
+        var isTypedArray = Object.prototype.toString.call( value ) === '[object Float32Array]';
```
Comment (stephomi) [inline sources/osg/Uniform.js]: value instanceof Float32Array ?

File: sources/osgShader/Compiler.js [hunk 3]
```diff
@@ -945,6 +955,34 @@ define( [
...
+            // Attribute 0 Must Be vertex
+            // perf warning in console otherwiser in opengl Desktop
+            if ( func._text.length ) {
+                // sort in alphabetical order
+                // attr, unif, sample, varying
+                func._text.sort();
+                // now sort Attributes
+                // making sure Vertex is always coming first
+                var toShift = [];
+                for ( j = 0; j < func._text.length && func._text[ 0 ].indexOf( 'Vertex' ) === -1; j++ ) {
```
Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: it's hard to read 
j < func._text.length && func._text[ 0 ].indexOf( 'Vertex' )
could simplify the logic or split into multiline

File: sources/osgAnimation/RigTransformHardware.js [hunk 4]
```diff
@@ -15,84 +19,74 @@ define( [
     function getShader( nbVec4 ) {
```
Comment (cedricpinson) [inline sources/osgAnimation/RigTransformHardware.js]: Do we still to keep this shader ?

Comment (Kuranes) [inline sources/osgAnimation/RigTransformHardware.js]: is for debug purpose ? 
To remove when  validated the remove of the "timeout" thingy
(as I couldn't reproduce here the bug that made that timeout necessary)

File: sources/osgShader/Compiler.js (hunk 3 above)
Comment (Kuranes) [inline sources/osgShader/Compiler.js]: Can add a "\n" after the "&&"  ?
Or you want to move a test inside the loop with a break ?

Comment (cedricpinson) [inline sources/osgShader/Compiler.js]: it's better out of the loop

File: sources/osgAnimation/RigTransformHardware.js (hunk 4 above)
Comment (cedricpinson) [inline sources/osgAnimation/RigTransformHardware.js]: well it was as first implementation to test without shader. If we dont need it anymore, we can remove it.

**Pull request merged**